### PR TITLE
A bit more garbage

### PR DESCRIPTION
**Pull request by sbird**
Continuing in our series of improving the garbage collection - this is a series with a few more tweaks. The differences are actually rather slight. The main commit is 9ba9d25, which adds a new garbage collection algorithm which is O(n) when compacting the slots, and preserves the (peano) order of the particles. This also lets the non-slot-compacting part avoid changing the slot array entirely, but because more data is moved, the base gc is slower. On my machine this works out to exactly the same time taken doing GC. The slot gc might be faster but with the default parameters it doesn't run very often. 
I also moved the peano sort into slotmanager and added a slot sort after it.

The main advantage is that treebuild is 20% faster on average, because the ordered particles make it easier for the optimizations to kick in. The better ordering makes no difference to the treewalk. Overall this made surprisingly little difference, which may suggest we have optimized as much as we can.

### CODE REVIEW

File: slotsmanager.c [hunk 1]
```diff
@@ -301,28 +295,32 @@ slots_gc_slots()
...
+        if(SlotsManager->info[ptype].garbage >= defrag)
+            disabled = 0;
     }
 
+    if(!disabled) {
+        slots_gc_mark();
+
+        for(ptype = 0; ptype < 6; ptype++) {
+            int defrag = 1 + defrag_frac * SlotsManager->info[ptype].size;
+            if(SlotsManager->info[ptype].garbage < defrag)
```
Comment (rainwoodman) [inline slotsmanager.c]: It looks strange this is done twice.

File: slotsmanager.c [hunk 2]
```diff
@@ -210,62 +232,34 @@ static int
 slots_gc_sweep(int ptype)
```
Comment (rainwoodman) [inline slotsmanager.c]: sweep removes

File: slotsmanager.c [hunk 3]
```diff
@@ -210,62 +232,34 @@ static int
...
+    for(i = 0; i < used; i++)
+        if(BASESLOT_PI(i,ptype)->IsGarbage)
+            slotBufcnt[sc++] = i;
+    slotBufcnt[sc] = used;
+    for(i = 0; i < sc; i++) {
+        int src = slotBufcnt[i]+1;
+        int dest = slotBufcnt[i] - i;
+        int nmove = slotBufcnt[i+1] - slotBufcnt[i]-1;
+//             message(1,"ptype = %d i = %d, PI = %d-> %d, nm=%d\n",ptype, i, src, dest, nmove);
+        /* TODO: in principle we can track this change and modify the tree nodes;
```
Comment (rainwoodman) [inline slotsmanager.c]: So the variable slotBufcnt is the 'permutation' array between the old PI and the new PI of the slot -- could you write a comment 'slotBufcnt[i] = j where i is XXX and j is XXX'? I don't understand track changes and modifying the tree nodes: tree doesn't care about PI -- it only care about the base particle index.

File: slotsmanager.c [hunk 4]
```diff
@@ -141,32 +141,41 @@ slots_gc(void)
...
+        for(i = 0; i < NumPart; i++)
+            if(P[i].IsGarbage) {
+                /* mark the particle for removal in base slot.*/
+                partBufcnt[pc++] = i;
+            }
+
+        /*Set final elements*/
+        partBufcnt[pc] = NumPart;
+
+        /*Compactify the P array: this invalidates the ReverseLink.*/
```
Comment (rainwoodman) [inline slotsmanager.c]: .. invalidates the ReverseLink causing ReverseLink only valid within slots_gc_slots.

File: slotsmanager.c [hunk 5]
```diff
@@ -346,7 +409,7 @@ slots_reserve(int atleast[6])
     if(SlotsManager->Base == NULL)
         SlotsManager->Base = (char*) mymalloc("SlotsBase", 0);
 
-    /* FIXME: change 0.005 to a parameter. The expericence is 
+    /* FIXME: change 0.005 to a parameter. The experience is
```
Comment (rainwoodman) [inline slotsmanager.c]: 0.01, not 0.005..

File: slotsmanager.c [hunk 6]
```diff
@@ -434,7 +497,10 @@ slots_mark_garbage(int i)
     P[i].IsGarbage = 1;
     if(SLOTS_ENABLED(P[i].Type)) {
         BASESLOT(i)->IsGarbage = 1;
+        atomic_add_and_fetch(&SlotsManager->info[P[i].Type].garbage, 1);
     }
+    /*Increment the counter*/
+    atomic_add_and_fetch(&SlotsManager->garbage, 1);
```
Comment (rainwoodman) [inline slotsmanager.c]: Why count for garbage here?

It's usually more efficient to avoid atomic operation and count the number of garbages if we have to run though the full particle storage anyways during GC.

If all we need is whether there is any garbage, then we don't need an atomic write -- just set has_garbage to 1.

Comment (rainwoodman) [reply slotsmanager.c]: In either case you do as many as +1 operations -- the atomic version is hell a lot worse than the non-atomic version.

File: slotsmanager.c (hunk 3 above)
Comment (sbird) [reply slotsmanager.c]: The TODO comment was left over from an earlier version - I will remove it.

File: slotsmanager.c (hunk 2 above)
Comment (sbird) [reply slotsmanager.c]: Fixed.

File: slotsmanager.c (hunk 6 above)
Comment (sbird) [reply slotsmanager.c]: The per-type atomics are there because we need to know if more than 10% of the slot is garbage to skip doing a gc.  The only other way to know that is to loop over the BASESLOT array. If we need to do that even if we are skipping the slot (which in practice we always are), gc becomes 50% slower and that completely dominates over the extra locking cost.

mark_slot_garbage is not called very often except during exchange, and the loop there is not threaded. Also when it is called there is another atomic access just before in fork, which will serialise.

 I can pretty easily avoid the overall garbage count though, if you think that is worth it. We could also just comment out the slots_gc_slots() - it is never called in practice because every PM step domain_decompose_full takes care of the garbage. What do you think?

Comment (rainwoodman) [reply slotsmanager.c]: Indeed during tree walk, not often. Though I recall during exchange we might call this quite often -- since we mark all outgoing particles as garbage there. Though I am not sure how bad a performance hit it really will be since the exchange is likely a lot more expensive anyways. So I agree with you now on keeping the atomic locks.

Avoid having an overall garbage count reduces the amount of book keeping between GC cycles -- which sounds like a cleaner design -- remember the relief when we get rid of the book keeping code for TotNumPart and TimBinCount? However it appears that we need some of these to keep track if it is necessary to do a GC, so let's keep them. Could you change the name of variable to ngarbage to indicate it is the number of garbage and add a comment in the struct to say it is maintained by using mark() during cycles?

I am a bit supprised slots_gc_slots is no longer used. I need to digest this.

Comment (rainwoodman) [reply slotsmanager.c]: The exchange algorithm uses gc to purge the out-going particles before receiving. So you can't really remove the routine -- otherwise we will run out of the slots quickly -- in cases where multiple iterations of exchange is required. 

Is the difference betweens slots_gc and slots_gc_sorted only that the latter sorts the particles before GC? If so perhaps it is cleaner to have a slots_sort_base(), and call both functions rather than having a larger function that does both.

Comment (sbird) [reply slotsmanager.c]: We do call it often during exchange - but that loop is single-threaded, so there is zero contention.

slots_gc_sorted() does a sort of the particles, moving garbage to the end. It sorts the base array and then it sorts the slots array. It doesn't use the O(n) garbage algorithm in slots_gc() at all, because it reorders rather than just compacting. I think I will add a comment to this effect.

So slots_gc_slots() is no longer called often in normal circumstances because to be called we need > 10% of the particles exchanged between two PM timesteps, and the displacement constraint  means that this doesn't happen often. 

You make a good point though that under low memory conditions with multiple exchange iterations we do need it - the slots could use up all the memory otherwise. In fact, I will make it explicit that it always runs when multiple iterations happen, even if it would compact by < 10% of the slots.

**Comment by sbird** [comment]
Ok. This new version took me far too long to figure out, but it avoids the counters. Instead it gc's the slots in an exchange only when we are short on memory. It also loops just once, so that it avoids allocating memory and is somewhat faster.

File: run.c [hunk 7]
```diff
@@ -186,18 +186,20 @@ void run(void)
...
 
-            /* the accel may have created garbage -- collect them before checkpointing! */
-            force_tree_free();
-            slots_gc();
-            force_tree_rebuild();
+            /* The accel may have created garbage -- collect them before checkpointing!
+             * Tree will be auto-rebuilt if gc collected particles,
+             * but we should rebuild the active list.*/
+            int compact[6] = {0};
+            if(slots_gc(compact))
```
Comment (rainwoodman) [inline run.c]: `slots_gc(need_compact, &base_moved);`?

Comment (sbird) [reply run.c]: What does the base_moved argument do?

File: slotsmanager.c [hunk 8]
```diff
@@ -127,46 +130,90 @@ slots_gc(void)
...
     tree_invalid |= slots_gc_base();
-    tree_invalid |= slots_gc_slots();
+    tree_invalid |= slots_gc_slots(compact_slots);
 
     MPI_Allreduce(MPI_IN_PLACE, &tree_invalid, 1, MPI_INT, MPI_SUM, MPI_COMM_WORLD);
 
-    EISlotsAfterGC event = {0};
+    EISlotsAfterGC event = {{tree_invalid}};
 
     event_emit(&EventSlotsAfterGC, (EIBase*) &event);
```
Comment (rainwoodman) [inline slotsmanager.c]: What if we add rebuild_activelist as an event handler?

File: slotsmanager.c [hunk 9]
```diff
@@ -127,46 +130,90 @@ slots_gc(void)
...
     return tree_invalid;
 }
 
+
+#define GARBAGE(i, ptype) (ptype >= 0 ? BASESLOT_PI(i,ptype)->IsGarbage : P[i].IsGarbage)
+#define PART(i, ptype) (ptype >= 0 ? (void *) BASESLOT_PI(i, ptype) : (void *) &P[i])
+
+/*Compaction algorithm*/
+static int
+slots_gc_compact(int used, int ptype, size_t size)
```
Comment (rainwoodman) [inline slotsmanager.c]: It feels a bit strange using ptype < 0 for the base slots -- I sort of felt baseslots are different from secondary slots. Perhaps we shall have both inheriting from GCFriendly, which is a 4 bytes or 8 byte struct with IsGarbage flag?

File: run.c (hunk 7 above)
Comment (rainwoodman) [reply run.c]: as a named return value ..

Comment (rainwoodman) [reply run.c]: But it sounds like rebuild_activelist can be an eventhandler registered by the timebin manager module. Then you get get rid of the call to rebuild_activelist here.

File: slotsmanager.c (hunk 8 above)
Comment (sbird) [reply slotsmanager.c]: We could - I'd rather not have it in a whole handler of it's own. The point of adding it is to ensure that when the tree gets rebuilt, the active list also gets rebuilt, so we should have them in the same function. But adding it to the existing handler would introduce a cross-module dependency of forcetree.c on timestep.c, and I'm not sure this is worth it, since at the moment this is only done in one place. What do you think?

Comment (sbird) [reply slotsmanager.c]: Hmm, no, on second thoughts lets have it in its own handler.

**Comment by rainwoodman** [comment]
Do we trigger the GC collectively on all ranks? I think we shall probably do that because if one rank hits GC, others will have to wait anyways.

File: slotsmanager.c (hunk 8 above)
Comment (sbird) [reply slotsmanager.c]: If it is in its own handler, then the active list will also be rebuilt after slot_gc in domain_exchange: so we would remove the rebuild_activelist call in the mainloop. But then we have a problem, which is that if the exchange does not generate garbage we would still need to rebuild the active list.

Also hiding something that is really a mainloop action in an event handler makes things harder to follow: working out what happens in event handlers is a bit tricky and over-using them can lead to crazy spaghetti code (parts of KDE in particular suffer from this). So I think on balance I prefer not to add it as an event handler.

File: slotsmanager.c (hunk 9 above)
Comment (rainwoodman) [reply slotsmanager.c]: This is better to be left on a future PR though.

File: slotsmanager.c (hunk 8 above)
Comment (sbird) [reply slotsmanager.c]: The better fix might be to move force_tree_rebuild back into the main loop and send the event handler back to being an error.

File: run.c (hunk 7 above)
Comment (sbird) [reply run.c]: I can do this instead of the event handler, but returning the number of items moved is a fairly standard C library convention, so I quite like it like this.

**Comment by sbird** [comment]
The base is collective, but the slot compaction is not - I will fix that.

File: slotsmanager.c (hunk 9 above)
Comment (sbird) [reply slotsmanager.c]: Yes - let's leave it like this for now and make it a future PR if we find ourselves doing this in more than one place.

File: slotsmanager.c (hunk 8 above)
Comment (sbird) [reply slotsmanager.c]: I have done this - the event handler is now gone.

Comment (rainwoodman) [reply slotsmanager.c]: Sure. The event handlers are to decouple interaction between modules -- the missing part is probably a diagram that charts who emits what and who listen to what -- in UI toolkits it's been too frequently abused as a way to avoid the overhead cost of subclassing -- I suppose that's what makes you dislike Events.

I agree that if the mainloop already depend on all modules then there is no need to use it. I thought it might be cleaner to keep all forcetree logic in forcetree and timebin logic in timebin -- where a very thin mainloop does almost nothing. So another approach is to say the event handler has an argument about number of baseslots moved. Then we can decide if to rebuild the tree in the force tree handler, and decide to rebuild the activelist in the timebin handler.

File: slotsmanager.c [hunk 10]
```diff
@@ -335,6 +346,72 @@ slots_gc_slots()
...
+            SlotsManager->info[ptype].size--;
+        }
+        slots_gc_collect(ptype);
+    }
+#ifdef DEBUG
+    slots_check_id_consistency();
+#endif
+
+    /*Rebuild the tree if still allocated*/
+    EISlotsAfterGC event = {{1}};
```
Comment (rainwoodman) [inline slotsmanager.c]: If we no longer need the event -- then shall remove it as well.. the comment looked out dated.

File: exchange.c [hunk 11]
```diff
@@ -165,19 +166,27 @@ static int domain_exchange_once(int (*layoutfunc)(int p), ExchangePlan * plan)
...
 
     ta_free(toGoPtr);
     walltime_measure("/Domain/exchange/makebuf");
 
-    /* now remove the garbage particles because they have already been copied.
-     * eventually we want to fill in the garbage gap or defer the gc, because it breaks the tree.
-     * invariance . */
+    /*Find which slots to gc*/
+    int compact[6] = {0};
+    for(ptype = 0; ptype < 6; ptype++) {
```
Comment (rainwoodman) [inline exchange.c]: Shall we make this a function, e.g. part of the plan?

File: slotsmanager.c (hunk 8 above)
Comment (sbird) [reply slotsmanager.c]: I don't dislike events per se - I think the current post-fork event is an excellent use of them, because it does remove cross-module dependencies and it describes something that could conceptually happen at any time in the code, so maps well onto what I understand an event to be. 

I just don't like over-using them: I saw once a mainloop which was implemented by several events chaining themselves together, which was just really hard to follow. Because they are attached at runtime it can become hard to follow which event handlers are actually running and debugging becomes tricky. I fixed a bug in nepomuk once where an event was being disconnected under peculiar circumstances deep within some other handler - it took me forever to figure it out, because I didn't have a way to find the function disconnecting it.

So if the main loop depends on all modules then it should explicitly state that in the header files - hiding the dependency in an event is just confusing.

One good reason for events is optional extensibility - I can see several of our feedback modules being implemented that way eventually. For example, SPH_GRAD_RHO could be replaced with an event handler registered by the H2 star formation criterion.

File: slotsmanager.c (hunk 10 above)
Comment (sbird) [reply slotsmanager.c]: Ok, done.

File: exchange.c (hunk 11 above)
Comment (sbird) [reply exchange.c]: Done

**Comment by sbird** [comment]
Travis is red but this seems to be a transient failure to install a dependent module (bigfile). Shall we merge?

**Comment by rainwoodman** [comment]
Yes. Please merge.

The compilation error is because the new mpich requires the compiler
packages from conda in order to build binaries (because it was built with
the compilers from conda).
